### PR TITLE
Flowchart: Apply `nodeSpacing`/`rankSpacing` config to subgraphs

### LINKED ISSUE #3258: Flowchart node spacing doesn't apply when using subgraph
I want to render some flowchart diagrams in a compact way to be readable on mobile, and found out that you can specify some distances between nodes using **nodeSpacing** (and **rankSpacing**), but it seems that when you start defining the nodes in a subgraph, it won't apply anymore.

1. Scenario 1 - nodes without subgraphs:
a) node spacing 10:

```
%%{init:{'flowchart':{'nodeSpacing': 10}}}%%

flowchart LR

A --> B
C --> D
```

[](https://mermaid.live/edit#pako:eNo9jLEKwjAURX8lPAhdWtA1g6DWzcmub3kksQ02LyWmiIT8u1HU7Rzu4WbQwVhQIGV27JLKzXUODz1RTE0VruuwkHY8NkpsN6UUKZGR_5U4X96-F123Ewfk4wd6aMHb6MmZep6RhUBIk_UWQVU0FG8IyKV262Io2ZNxKURQV5rvtgVaUxierEGluNpf1DsaI_lvVV5lRD-m)

b) node spacing 50:

```
%%{init:{'flowchart':{'nodeSpacing': 50}}}%%

flowchart LR

A --> B
C --> D
```

[](https://mermaid.live/edit#pako:eNo9jLEKwjAURX8lPAhdWnBxySCodXOy61seSWyDzUuJKSIh_24UdTuHe7gZdDAWFEiZHbukcnOdw0NPFFNThes6LKQdj40S200pRUpk5H8lzpe370XX7cQB-fiBHlrwNnpypp5nZCEQ0mS9RVAVDcUbAnKp3boYSvZkXAoR1JXmu22B1hSGJ2tQKa72F_WOxkj-W5UXZ4Q_qg)

Renders well, the nodes are properly spaced.

2. Scenario 2, node within subgraphs:
a) node spacing 10:

```
%%{init:{'flowchart':{'nodeSpacing': 10, 'rankSpacing':'10'}}}%%

flowchart LR

X --> Y
subgraph X
	direction LR
	A
	C
end

subgraph Y
	direction LR
	B
	D
end
```

[](https://mermaid.live/edit#pako:eNptj0GLwjAQhf9KGCi5VNBrDoK77s2TXlrIZUxGG7STEqeIlPx3U3dXWNjL472Zb3jMBC56AgNVNQUOYiZ9usa76zCJLoHL9jCgC3zWRq2WtdIJ-fIe6dVS55yryrLl96Xa7efcqMVirdrZ3sbjOeHQqaYk8SGRkxD5G5TNLJ-Wif0fuP0P_phl-4Khhp5Sj8GXDybLSlmQjnqyYIr1mC4WLOfCjYNHoS8fJCYwJ7zeqAYcJR4e7MBIGukX2gYs9f0PlZ-g7WYW)

b) node spacing 50:

```
%%{init:{'flowchart':{'nodeSpacing': 50, 'rankSpacing':'10'}}}%%

flowchart LR

X --> Y

subgraph X
	direction LR
	A
	C
end

subgraph Y
	direction LR
	B
	D
end
```
[](https://mermaid.live/edit#pako:eNptjz-LwzAMxb-KEQQvKbTDLR4K_bd1ui4teFFttTFp5OAqHEfwd6_TuxYObnm8J_3EQyO46AkMVNUYOIgZ9eUWv1yDSXQJXLaHHl3gqzbqY14rnZDb90gv5jrnXFWWLb8v1f5zykc1my3VabL34XxN2DfqWJL4kMhJiPwDymqSjWVi_wc-_QevJ9k-Yaiho9Rh8OWD0bJSFqShjiyYYj2m1oLlXLih9yi080FiAnPB251qwEHi4ZsdGEkDvaBtwFLf_VL5AaTBZho)

Node spacing doesn't apply anymore when these nodes are inside subgraph, the vertical distance between them remains the same.

Expected: To be able to apply node spacing inside a subgraph as well.
Environment: Latest Mermaid Live Editor (9.1.3), Obsidian 0.5.18 (don't know what Mermaid version they're using, but mentioning it here since there I found it first).

### PR DESCRIPTION
**Pull request by rowanfr**
## :bookmark_tabs: Summary

This commit resolves the issues with subgraph configuration, specifically for `nodeSpacing` and `rankSpacing`. This commit additionally adds an example graph to the `flowchart.html` to demonstrate this resolution.

Resolves #3258

## :straight_ruler: Design Decisions

This is a fairly simple modification that applies the configuration held in the label for the overall graph to the configuration for the cluster/subgraph in order to ensure they have matching configurations. This is done right before `recursiveRender` begins to render each cluster. The reason this needs to be done is that the label for the subgraphs do not contain the correct `nodeSpacing` and `rankSpacing` based on the mermaid configuration. Therefore if one overwrites the graph label right before rendering it resolves the mismatch of configurations.

Some potential improvements for this commit would be to allow subgraphs to have there own customization labels. This would be fairly trivial (presumably) by just replacing `graph.graph()` with whatever settings you want from whatever configurations you want.

### :clipboard: Tasks

Make sure you

- [X] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
:computer: have added necessary unit/e2e tests. (Not necessary to my knowledge though I'm open to alternative opinions)
:notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/development.md#3-update-documentation) is used for all new features. (Not necessary to my knowledge as this is a bugfix)
- [X] :bookmark: targeted `develop` branch

### CODE REVIEW

File: demos/flowchart.html [hunk 1]
```diff
@@ -1540,6 +1540,30 @@ <h3>flowchart</h3>
     </pre>
     <hr />
 
+    <pre class="mermaid">
+      ---
+      title: Subgraph nodeSpacing and rankSpacing example
+      ---
+      %%{init:{'flowchart':{'nodeSpacing': 1, 'rankSpacing':'1'}}}%%
```
Comment (nirname) [inline demos/flowchart.html:1547]: Avoid directives configuration, which is obsolete, use frontmatter
```suggestion
      ---
      title: Subgraph nodeSpacing and rankSpacing example
      config:
        flowchart:
          nodeSpacing: 1
          rankSpacing: 1
      ---
```

**Review by nirname — CHANGES_REQUESTED** [review]
Hi! Thanks for addressing the issue! Deployment failed, we will have to investigate why, but regardless of this there are some minor comments.

**Comment by rowanfr** [comment]
@nirname Thanks for the suggested change in syntax, I didn't even know that was available

**Comment by sidharthv96** [comment]
@rowanfr there are multiple failing tests. 

Flowchart-v2-57-handle-nested-subgraphs-with-outgoing-links-4 diff
Flowchart-v2-66-More-nested-subgraph-cases-(TB) diff
Flowchart-v2-67-More-nested-subgraph-cases-(RL) diff
Flowchart-v2-68-More-nested-subgraph-cases-(BT) diff
Flowchart-v2-69-More-nested-subgraph-cases-(LR) diff
Flowchart-v2-2050-handling-of-different-rendering-direction-in-subgraphs diff
Flowchart-v2-Subgraph-title-margins-Should-render-subgraphs-with-title-margins-and-edge-labels diff
Flowchart-v2-Subgraph-title-margins-Should-render-subgraphs-with-title-margins-set-(LR)-and-htmlLabels-set-to-false diff
Flowchart-v2-Subgraph-title-margins-Should-render-subgraphs-with-title-margins-set-(LR) diff
Flowchart-v2-Subgraph-title-margins-Should-render-subgraphs-with-title-margins-set-(TD) diff
Graph-30-Possibility-to-style-text-color-of-nodes-and-subgraphs-as-well-as-apply-classes-to-subgraphs diff
Graph-64-fontawesome-icons-in-edge-labels diff
Flowchart-v2-63-title-on-subgraphs-should-be-themable diff
Flowchart-v2-52-handle-nested-subgraphs-in-several-levels diff
Flowchart-v2-53-handle-nested-subgraphs-with-edges-in-and-out diff
Flowchart-v2-57-handle-nested-subgraphs-with-outgoing-links-2 diff
Flowchart-v2-55-handle-nested-subgraphs-with-outgoing-links-2 diff

**Comment by rowanfr** [comment]
@sidharthv96 Could you tell me how or where you encountered the errors that your showing? I'm using the dev server to get output data and they all seem to be resolving as expected.
image

**Comment by sidharthv96** [comment]
It was observed after the run of the E2E test that was triggered by the pull request event (not the push event).

**Comment by rowanfr** [comment]
After some research the e2e test failing only doesn't seem to occur locally outside of randomly when changing the window or running other applications, however Cypress errors for `rendering/flowchart-v2.spec.js` and `rendering/flowchart.spec.js` don't occur locally on Firefox Nightly or Electron after testing. I think this means that the tests are configured to take snapshots too early, but I'm still investigating it.

**Comment by rowanfr** [comment]
For the examples above all seem to have the links and graphs shift position slightly. This seems to be an browser dependent issue as I can test locally on electron it repeatedly produce successful graphs for Cypress, firefox nightly seems to randomly produce successful graphs while edge and chrome do not. As I've stated earlier I'll look into this a bit further but I think this has to do with browser rendering diff, which I don't have the ability to address

**Comment by sidharthv96** [comment]
Hmm, The E2E tests are working fine on other PRs, so it means some change in this one is triggering the browser inconsistency, and it's failing consistently as well, not flaky. 

Maybe your change has some other effects downstream that's causing the issue. We'll have to investigate further. 

@aloisklink @nirname @Yokozuna59 do you folks have any pointers?

File: packages/mermaid/src/dagre-wrapper/index.js [hunk 2]
```diff
@@ -54,6 +54,9 @@ const recursiveRender = async (_elem, graph, diagramtype, id, parentCluster, sit
       if (node && node.clusterNode) {
         // const children = graph.children(v);
         log.info('Cluster identified', v, node.width, graph.node(v));
+        // node.graph.setGraph applies the graph configurations such as nodeSpacing to subgraphs as without this the default values would be used
+        // We override the `rankdir` variable to allow for the subgraph's direction to override the graphs direction in the subgraph
+        node.graph.setGraph({ ...graph.graph(), rankdir: node.graph.graph().rankdir });
```
Comment (aloisklink) [inline packages/mermaid/src/dagre-wrapper/index.js:59]: I think I've found the issue that was causing the screenshots of the E2E tests to fail.

The issue is that we were updating `marginy` for subgraphs to use their parent value, and that was causing the positioning to shift very slightly.

If we instead only copy over the `ranksep` (i.e. the [`rankSpacing`](https://mermaid.js.org/config/schema-docs/config-defs-flowchart-diagram-config.html#rankspacing)) and the `nodesep` (i.e. the [`nodeSpacing`](https://mermaid.js.org/config/schema-docs/config-defs-flowchart-diagram-config.html#nodespacing)), it seems to work fine.

In the real world, however, it should would slightly modify the subgraphs of anybody that has set `rankSpacing`/`nodeSpacing`, so IMO, this should be in a MINOR release, not a PATCH release.

```suggestion
        const { ranksep, nodesep } = graph.graph();
        node.graph.setGraph({
          ...node.graph.graph(),
          ranksep,
          nodesep,
        });
```

File: demos/flowchart.html [hunk 3]
```diff
@@ -1540,6 +1540,33 @@ <h3>flowchart</h3>
...
+        direction LR
+        A
+        C
+      end
+      
+      subgraph Y
+        direction LR
+        B
+        D
+      end
```
Comment (aloisklink) [inline demos/flowchart.html:1566]: Awesome for adding a demo!

Can you also add this demo (or another diagram) to the [`cypress/integration/rendering/flowchart-v2.spec.js`](https://github.com/mermaid-js/mermaid/blob/d6e738ac4cf2e208326b0ca3a8702d940cb846ec/cypress/integration/rendering/flowchart-v2.spec.js) file? This is just so we get some automated screenshots of the new feature you added :)

It should look something like:

```ts
  it('Should render subgraphs with different nodeSpacing', () => {
    imgSnapshotTest(`---
title: Subgraph nodeSpacing and rankSpacing example
---
flowchart LR
  X --> Y
  subgraph X
    direction LR
    A
    C
  end
  subgraph Y
    direction LR
    B
    D
  end
`,
      { flowchart: { nodeSpacing: 1, rankSpacing: 1 } }
    );
  });
```

**Review by aloisklink — CHANGES_REQUESTED** [review]
I think I've spotted the issue that was causing the E2E tests to fail, we just need to slightly modify the code so that it doesn't copy the parent graph's margin.

It would also be nice to have a very basic E2E test in the [cypress/integration/rendering/flowchart-v2.spec.js](https://github.com/mermaid-js/mermaid/blob/d6e738ac4cf2e208326b0ca3a8702d940cb846ec/cypress/integration/rendering/flowchart-v2.spec.js) file (e.g. your demo diagram), just so we get automated tests so we know not to break your feature!

Other than that, this PR looks good to me.

**Comment by rowanfr** [comment]
@aloisklink Thanks for spotting the issue that was causing the E2E tests to fail. That problem genuinely had me completely stumped for months. I've added the E2E testing for subgraph configuration as you suggested

**Review by aloisklink — APPROVED** [review]
Looks awesome to me now! Btw, I've slightly adjusted the PR title, let me know if you don't like it and want to change it back (or to something else).

> Thanks for spotting the issue that was causing the E2E tests to fail. That problem genuinely had me completely stumped for months.

Yep, the `dagre-wrapper/index.js` file is really really hard to read. I really want to look into converting it to TypeScript and adding at least some documentation to it, but I think we might also need to fix some types in a third-party library (and that library is also very undocumented), so it would be a lot of work :(

**Comment by rowanfr** [comment]
The new title looks great and is more accurate to the solution, the earlier title was generally what was desired so the change to it makes sense. 

As for `dagre-wrapper/index.js` I agree that unraveling that back to third party libraries that eventually rely on `graphlib` graph labels for rendering logic was fairly difficult and TypeScript alongside further documentation of the third party libraries would go a long way to help

**Comment by rowanfr** [comment]
The E2E tests are failing due to some separate git graph tests that were failing due to randomized commit IDs as detailed in the issue #5343 and the subsequent PR #5344. If we merge those changes minor changes the E2E testing for this PR should be resolved.

**Comment by rowanfr** [comment]
@nirname could you check over the changes you've requested to see if they've been completed and then approve the PR if you feel confident?

**Comment by Stevio54** [comment]
Is there an update to this pull request, I am having the same issues and would love a solution :)

**Comment by rowanfr** [comment]
@Stevio54 It's feature complete, it just needs to be merged

**Pull request merged**